lib/restforce.rb

```diff
@@ -45,6 +45,24 @@ module Tooling
   UnauthorizedError   = Class.new(Error)
   APIVersionError     = Class.new(Error)
 
+  # Inherit from Faraday::Error::ResourceNotFound for backwards-compatibility
+  # Consumers of this library that rescue and handle Faraday::Error::ResourceNotFound
+  # can continue to do so.
+  NotFoundError       = Class.new(Faraday::Error::ResourceNotFound)
+
+  # Inherit from Faraday::Error::ClientError for backwards-compatibility
+  # Consumers of this library that rescue and handle Faraday::Error::ClientError
+  # can continue to do so.
+  ResponseError       = Class.new(Faraday::Error::ClientError)
+  MatchesMultipleError= Class.new(ResponseError)
+  EntityTooLargeError = Class.new(ResponseError)
+
+  module ErrorCode
+    def self.const_missing(constant_name)
```

Comment:
I have a natural aversion to `const_missing` - maybe we could use `ResponseError` and just attach an instance variable to it exposing the error code?

Reply:
What's your aversion to `const_missing`? Is there a danger to it that we could mitigate?

Exposing the error code with an instance variable is fine doesn't really achieve what the constants do. With separate classes for each of Salesforce's error codes:
 1. You can `rescue` from specific error codes without resorting to helper methods or metaprogramming
 2. It's friendlier to Sentry and Airbrake's fingerprinting algorithms. (Sentry uses exception class + backtrace (I believe) to fingerprint errors and, when they are all have the same class name (before this PR), Sentry lumps together exceptions caused by _very different_ Salesforce errors. 😬)

* * *

We could move the guts of `const_missing` to `exception_class_for_error_code`. _That_ method could end like this:

```ruby
if Restforce::ErrorCode.const_defined?(constant_name)
  Restforce::ErrorCode.const_get(constant_name)
else
  Restforce::ErrorCode.const_set(constant_name, Class.new(Restforce::ResponseError))
end
```

Would that be better?